Feature download category metadata

Amejia481 opened the pull request:
Close #274

Amejia481 committed:
Download category, category options & category combinations and all its dependencies

Amejia481 committed:
Merge branch 'development' of https://github.com/EyeSeeTea/dhis2-android-sdk into feature-download_category_metadata

ifoche committed:
merged with development

ifoche committed:
rename CallEndpoint by EndpointCall according to the standard

ifoche commented on core/src/androidTest/assets/categories.json:
```diff
@@ -0,0 +1,1026 @@
+{
```
@Amejia481 I think for mocked json it would be better to have small examples instead of big jsons with too much information that we don't use in the tests. Unless you chose such a big example for a concrete reason...could you please change it to something containing what you actually need? Thank you!

Amejia481 replied:
Sure!
Thanks for the feedback!

ifoche commented on core/src/androidTest/assets/categories_combo.json:
```diff
@@ -0,0 +1,4462 @@
+{
```
Same comment than before. If you don't need this amount of information for your test, I would cut it to something more easy to maintain

Amejia481 replied:
Sure!
Apart from connectedAndroidTest gradle task and 7 PMD rule violations, 
are there any Unit tests or Android Tests falling?

Amejia481 replied:
Ok, after the merge with the pull request #280, some tests were failing.
I'm on it!

ifoche commented on core/src/androidTest/java/org/hisp/dhis/android/core/category/CategoryComboCallEndpointRealShould.java:
```diff
@@ -0,0 +1,218 @@
+    // @Test
+    public void download_category_combos() throws Exception {
+
+        Response responseLogIn = d2.logIn(RealServerMother.user, RealServerMother.password).call();
```
Provided that you're using a real server and so crossing several layers of the application, I would call this test CategoryComboEnpointCallRealIntegrationShould (note the Integration between Real and Should).

ifoche committed:
change test name according to standard EndpointCall

ifoche reviewed (dismissed):
@Amejia481 I left some comments. Apart from that, there are some tests failing (see connectedAndroidTest gradle task result) and pmd throws also 7 PMD rule violations. Please, can you give it a look? Thanks!

Amejia481 commented:
**Checklist**
- [x] Fix falling tests after the merge with the pull request #280.
- [x] Check connectedAndroidTest Gradle task result.
- [x] Fix PMD rule violations.
- [x] Reducing the size of categories.json to just a small examples.
- [x] Reducing the size of categories_combo.json to just a small examples.
- [x] Create a branch to keep the handlers abstraction.
- [x] Rename test CategoryComboEndpointCallRealShould to CategoryComboEndpointCallRealIntegrationShould
- [x] Remove the handlers abstraction for the current branch.

ifoche commented:
@Amejia481 just for the record, as we talked in the team meeting, it would be great if you could create a separate branch and keep your current code there (to avoid losing the work you did regarding the handlers abstraction), and then on this PR just remove that abstraction. Thank you in advance!

Amejia481 commented:
Just for the record, I tried to narrow the scope that covers @SuppressWarnings, for the smallest scope possible, this way future new violations that are added to the source code will be shown.

Amejia481 committed:
- Fixed falling tests after the merge with the pull request #280.
- Checked connectedAndroidTest Gradle task result.
- Fixed PMD rule violations.
- Reducing the size of categories.json to just a small examples.
- Reducing the size of categories_combo.json to just a small examples.

Amejia481 commented:
I created the new brach to keep the handler abstraction features its name is [handler_abstractions ](https://github.com/EyeSeeTea/dhis2-android-sdk/tree/handler_abstractions)

Amejia481 committed:
#283
Remove the handlers abstraction for the current branch.
Rename test CategoryComboEndpointCallRealShould to CategoryComboEndpointCallRealIntegrationShould

Amejia481 commented:
The changes have been made please verify :+1:

ifoche committed:
merged with development

ifoche committed:
expose Category related stores from D2 and include into them the delete calls on logout

ifoche committed:
refactor comboService to the standard categoryComboService

ifoche committed:
standarize categoryxxx names

Amejia481 committed:
Fixing Timeout issue on instrumented test

Amejia481 committed:
Fixed
CategoryComboEndpointCallRealIntegrationShould
download_category_combos

Amejia481 commented on core/src/main/java/org/hisp/dhis/android/core/data/database/DbOpenHelper.java:
```diff
@@ -182,10 +182,6 @@
             " REFERENCES " + CategoryOptionComboModel.TABLE + " ("
             + CategoryOptionComboModel.Columns.UID + ") " +
             " ON DELETE CASCADE DEFERRABLE INITIALLY DEFERRED," +
-            " FOREIGN KEY (" + CategoryOptionComboLinkCategoryModel.Columns.CATEGORY + ") " +
```
Fixing foreign key issue on CategoryComboEndpointCallRealIntegrationShould test

Amejia481 committed:
Chaning the order in witch the categories
 are downloaded now they get dowonloaded
 before the programs

Amejia481 committed:
Merge remote-tracking branch 'origin/development' into feature-download_category_metadata

# Conflicts:
#	core/src/androidTest/java/org/hisp/dhis/android/core/LogoutCallMockIntegrationShould.java
#	core/src/main/java/org/hisp/dhis/android/core/D2.java

Amejia481 committed:
Merging with development

Amejia481 commented:
After merging with development, all the tests are passing  :grinning:

ifoche committed:
rename categories_combo by category_combos

ifoche committed:
add category models to assertion

ifoche committed:
reorder LogoutCallMockIntegrationShould tests to place tests first and aux private methods later

ifoche committed:
create a CategoryCallFactory and include a creator for cateogoryHandlers in HandlerFactory

ifoche committed:
remove cursor usage from category combos queries in tests

ifoche committed:
completely remove usage of cursors in CategorycomboEndpointCallRealIntegrationShould test

ifoche committed:
fix checkStyle errors

ifoche committed:
fix PMD error (make CatOptionComboStore local in D2)

ifoche committed:
dont delete db on test teardown

ifoche committed:
refactor CategoryComboLinkStoreShould test

ifoche committed:
remove teardown to delete db on categoryComboStoreShould test

ifoche committed:
refactor CategoryComboStoreShould test

ifoche committed:
create categoryCombos related factories

ifoche committed:
refactor CategoryOptionComboLinkCategoryStoreShould test

ifoche committed:
refactor CategoryOptionComboStoreShould

ifoche committed:
refactor CategoryOptionLinkStoreShould test

ifoche committed:
refactor CategoryOptionStoreShould test

ifoche committed:
refactor CategoryStoreShould

ifoche committed:
remove isNotValid function from D2 (function with 1 parameter vs negative condition)

ifoche committed:
change CategoryOptionComboLinkCategory by the standard CategoryOptionComboCategoryLink

ifoche committed:
revert test changes

ifoche committed:
remote unused interface for stores

ifoche committed:
rename db tables and table columns following the dhis2 naming conventions

ifoche reviewed (approved):
many thanks @Amejia481. We'll talk about the additions and refactors in our next meeting.
